skimage/_shared/gpu/convolve.py

```diff
@@ -0,0 +1,32 @@
+from scipy import ndimage as ndi
+import numpy as np
+import skimage._shared.gpu
+#setup the imports, attempt gputools, if not, fallback instantly
+#
+
+'''
+  mode_str = {"constant":"CLK_ADDRESS_CLAMP",
+                "wrap":"CLK_ADDRESS_REPEAT"}
+                "wrap":"CLK_ADDRESS_REPEAT",
+                "nearest":"CLK_ADDRESS_CLAMP_TO_EDGE",
+                "reflect":"CLK_ADDRESS_MIRRORED_REPEAT"}
+'''
+
+if(skimage._shared.gpu.use_gpu):
+    import skimage._vendor.gputools as gputools
+    def convolve(input,weights,output=None,mode='reflect',cval=0.0,origin=0):
+        if (mode == 'reflect') or (mode == 'constant' and cval == 0.0): #These are the only two conditions that gputools supports
+            print('using gpu!') #print for debugging
+            out = gputools.convolve(np.array(input),np.array(weights),mode=mode) 
+            # we need to do some type conversion here in order to maintain full compatibility
+            # as ndi autocasts to np arrays, but gputools does not
+            if output != None:
+                output = out
+        else:
+            print('falling back to scikit-ndi')
+            out = ndi.convolve(input,weights,output=output,mode=mode,cval=cval,origin=origin)
+
+        return out
+
+else:
+    convolve = ndi.convolve
```

Comment:
@scikit-image/core I want to draw your attention to this file, as well as the changes in [`skimage.filters._gabor.py`](https://github.com/scikit-image/scikit-image/pull/4215/files#diff-52301442ac0295262f831633e19dbec3), which only consist of changing the imported convolve version from `ndimage` to `skimage._shared.gpu`. (@Dalordish I would just make this `skimage._shared`, so the function is just `skimage._shared.convolve`.) imho it is worthwhile building this layer.

Reply:
It could be nice to implement a context manager as was done for the upcoming `scipy.fft` module so that users could run selected code blocks on GPU vs CPU with code similar to the following.

```Python
with skimage.backend('gputools'):
    out = gabor(image, 0.1)
```

or
```Python
with skimage.backend('cpu'):
    out = gabor(image, 0.1)
```

`scipy.fft` also includes `get_backend` and `set_backend` functions that can be used to set the backend globally. I think having something like that here would be nice as an alternative to only reading the global state once based on an environment variable.